migration/migration_test.go

```diff
@@ -0,0 +1,332 @@
+}
+
+func TestFailingMigration(t *testing.T) {
+	ms := map[int]Migration{
+		1: {
+			UseTxn: true,
+			Run: func(_ datastoreReaderWriter) error {
+				return fmt.Errorf("I failed")
+			},
+		},
+	}
+	m := newMigrator(ms, false)
+
+	v, _, err := m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+
+	err = m.Ensure()
+	require.Error(t, err)
+
+	v, _, err = m.getCurrentVersion()
+	require.NoError(t, err)
+	require.Equal(t, 0, v)
+}
+
+func TestRealDataBadger(t *testing.T) {
+	logger.SetDebugLogging()
+	_ = logger.SetLogLevel("badger", "error")
+	tmpDir := t.TempDir()
+	err := copyDir("testdata/badgerdump", tmpDir+"/badgerdump")
+	require.NoError(t, err)
+
+	opts := &badger.DefaultOptions
+	ds, err := badger.NewDatastore(tmpDir+"/badgerdump", opts)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, ds.Close()) }()
+
+	migrations := map[int]Migration{
+		1: V1MultitenancyMigration,
+	}
+	m := New(ds, migrations)
+	err = m.Ensure()
+	require.NoError(t, err)
+}
+
+func TestRealDataRemoteMongo(t *testing.T) {
+	t.SkipNow()
+	logger.SetDebugLogging()
+	mongoCtx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	mongoURI := ""
+	mongoDB := ""
+	ds, err := mongods.New(mongoCtx, mongoURI, mongoDB, mongods.WithCollName("migration_kvstore"), mongods.WithOpTimeout(time.Hour), mongods.WithTxnTimeout(time.Hour))
+	require.NoError(t, err)
+	defer func() { require.NoError(t, ds.Close()) }()
+
+	migrations := map[int]Migration{
+		1: V1MultitenancyMigration,
+	}
+	m := New(ds, migrations)
+	err = m.Ensure()
+	require.NoError(t, err)
+}
+
+func newMigrator(migrations map[int]Migration, empty bool) *Migrator {
+	ds := tests.NewTxMapDatastore()
+	m := New(ds, migrations)
+
+	if !empty {
+		_ = ds.Put(datastore.NewKey("foo"), []byte("bar"))
+	}
+
+	return m
+}
+
+func pre(t *testing.T, ds datastore.TxnDatastore, path string) {
+	t.Helper()
+	f, err := os.Open(path)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, f.Close()) }()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		parts := strings.SplitN(s.Text(), ",", 2)
+		err = ds.Put(datastore.NewKey(parts[0]), []byte(parts[1]))
+		require.NoError(t, err)
+	}
+}
+
+func post(t *testing.T, ds datastore.TxnDatastore, path string) {
+	t.Helper()
+
+	f, err := os.Open(path)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, f.Close()) }()
+
+	current := map[string][]byte{}
+	q := query.Query{}
+	res, err := ds.Query(q)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, res.Close()) }()
+	for r := range res.Next() {
+		require.NoError(t, r.Error)
+		current[r.Key] = r.Value
+	}
+
+	expected := map[string][]byte{}
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		parts := strings.SplitN(s.Text(), ",", 2)
+		expected[parts[0]] = []byte(parts[1])
+	}
+
+	require.Equal(t, len(expected), len(current))
+	for k1, v1 := range current {
+		v2, ok := expected[k1]
+		require.True(t, ok)
+		require.Equal(t, v2, v1)
+	}
+}
```

Comment:
`pre` is a function that reads a csv files with key,values as a initial state of a datastore for tests.
`post` is a function that asserts if a datastore has the same state as a csv file with key,values.